src/compiler/tsbuildPublic.ts

```diff
@@ -1459,80 +1506,182 @@ namespace ts {
-            };
+            oldestOutputFileTime = buildInfoTime;
+            oldestOutputFileName = buildInfoPath;
+            newestDeclarationFileContentChangedTime = buildInfo.program?.dtsChangeTime ? new Date(buildInfo.program.dtsChangeTime) : undefined;
         }
 
-        if (isOutOfDateWithInputs) {
-            return {
-                type: UpToDateStatusType.OutOfDateWithSelf,
-                outOfDateOutputFileName: oldestOutputFileName,
-                newerInputFileName: newestInputFileName
-            };
+        // Check input files
+        let newestInputFileName: string = undefined!;
+        let newestInputFileTime = minimumDate;
+        // Get timestamps of input files
+        for (const inputFile of project.fileNames) {
+            const inputTime = getModifiedTime(state, inputFile);
+            if (inputTime === missingFileModifiedTime) {
+                return {
+                    type: UpToDateStatusType.Unbuildable,
+                    reason: `${inputFile} does not exist`
+                };
+            }
+
+            // If an buildInfo is older than the newest input, we can stop checking
+            if (buildInfoTime && buildInfoTime < inputTime) {
+                return {
+                    type: UpToDateStatusType.OutOfDateWithSelf,
+                    outOfDateOutputFileName: buildInfoPath!,
+                    newerInputFileName: inputFile
+                };
+            }
+
+            if (inputTime > newestInputFileTime) {
+                newestInputFileName = inputFile;
+                newestInputFileTime = inputTime;
+            }
         }
-        else {
-            // Check tsconfig time
-            const configStatus = checkConfigFileUpToDateStatus(state, project.options.configFilePath!, oldestOutputFileTime, oldestOutputFileName);
-            if (configStatus) return configStatus;
-
-            // Check extended config time
-            const extendedConfigStatus = forEach(project.options.configFile!.extendedSourceFiles || emptyArray, configFile => checkConfigFileUpToDateStatus(state, configFile, oldestOutputFileTime, oldestOutputFileName));
-            if (extendedConfigStatus) return extendedConfigStatus;
-
-            // Check package file time
-            const dependentPackageFileStatus = forEach(
-                state.lastCachedPackageJsonLookups.get(resolvedPath) || emptyArray,
-                ([path]) => checkConfigFileUpToDateStatus(state, path, oldestOutputFileTime, oldestOutputFileName)
-            );
-            if (dependentPackageFileStatus) return dependentPackageFileStatus;
+
+        // Now see if all outputs are newer than the newest input
+        // Dont check output timestamps if we have buildinfo telling us output is uptodate
+        if (!buildInfoPath) {
+            // Collect the expected outputs of this project
+            const outputs = getAllProjectOutputs(project, !host.useCaseSensitiveFileNames());

```

Comment:
Output timestamps are old checks now done only  for non incremental projects as they dont have buildInfo to tell us their upto date ness. 
Also we dont need to get when declaration file changed last because that is needed only for composite projects which are incremental anyways.. (as those are the only that can be referenced and we need that d.ts time to decide if we build project again or just do pseudo update)